extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/S3Utils.java

```diff
@@ -168,6 +244,26 @@ public S3ObjectSummary next()
     };
   }
 
+
+  /**
+   * Create an {@link URI} from the given {@link S3ObjectSummary}. The result URI is composed as below.
+   *
+   * <pre>
+   * {@code s3://{BUCKET_NAME}/{OBJECT_KEY}}
+   * </pre>
+   */
+  public static URI summaryToUri(S3ObjectSummary object)
+  {
+    final String originalAuthority = object.getBucketName();
+    final String originalPath = object.getKey();
+    final String authority = originalAuthority.endsWith("/") ?
+                             originalAuthority.substring(0, originalAuthority.length() - 1) :
+                             originalAuthority;
+    final String path = originalPath.startsWith("/") ? originalPath.substring(1) : originalPath;
+
+    return URI.create(StringUtils.format("s3://%s/%s", authority, path));
```

Comment:
This is bad, because it won't encode funny characters in `path`. Imagine the `path` has a `?` in it. It needs to be URI-encoded, or else pulling the key out later won't work. The tricky characters are `/` (which you _don't_ want to encode) and `?`, `#`, and others (which you do).

`StringUtils.urlEncode` might help you here.

Alternatively, don't use URIs internally, instead use bucket/key pairs.

Reply:
How about adding this method to CloudObjectLocation:

```java
public URI toUri()
{
  // Encode path, except leave '/' characters unencoded
  return URI.create(StringUtils.format("s3://%s/%s", bucket, StringUtils.urlEncode(path).replace("%2F", "/"));
}
```

And using it everywhere that is doing this sort of concatenation today.

It won't handle weird, invalid `bucket` names but it's better than the simple concatenation happening now, and weird `path`s are more likely anyway. For extra credit you could include validation for the `bucket`, throwing an error if it's not valid (AWS, Google, etc all have rules for what's a valid bucket, you could do a loose superset of them).

Reply:
FYI this method in `S3Utils` was only called by `StaticS3FirehoseFactory`, not the new stuff, so I wasn't worrying about it too much because we presumably will remove it in a future release. That said, I went ahead and did the thing to fix it